pkg/maps/metricsmap/metricsmap.go

```diff
@@ -137,47 +137,134 @@ func (vs Values) Bytes() uint64 {
+			"Total dropped packets, tagged by drop reason and ingress/egress direction",
+			[]string{metrics.LabelDropReason, metrics.LabelDirection}, nil,
+		),
+		forwardCountDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "forward_count_total"),
+			"Total forwarded packets, tagged by ingress/egress direction",
+			[]string{metrics.LabelDirection}, nil,
+		),
+		forwardByteDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "forward_bytes_total"),
+			"Total forwarded bytes, tagged by ingress/egress direction",
+			[]string{metrics.LabelDirection}, nil,
+		),
+	}
+}
+
+func (mc *metricsmapCollector) Collect(ch chan<- prometheus.Metric) {
+
+	err := Metrics.IterateWithCallback(func(key *Key, values *Values) {
+		if key.IsDrop() {
+			labels := dropLabels{
+				direction: key.Direction(),
+				reason:    key.DropForwardReason(),
+			}
+
+			v := lookupOrCreateMetricsMapValues(mc.droppedMetricsMap, labels)
+			v.bytes += float64(values.Bytes())
+			v.count += float64(values.Count())
+			mc.droppedMetricsMap[labels] = v
+		} else {
+			labels := forwardLabels{
+				direction: key.Direction(),
+			}
+
+			v := lookupOrCreateMetricsMapValues(mc.forwardedMetricsMap, labels)
+			v.bytes += float64(values.Bytes())
+			v.count += float64(values.Count())
+			mc.forwardedMetricsMap[labels] = v
+		}
+	})
 	if err != nil {
-		log.WithError(err).Warn("Failed to update prometheus metrics")
+		log.WithError(err).Warn("Failed to read metrics from BPF map")
+		// do not update partial metrics
 		return
 	}
 
-	oldValue := metrics.GetCounterValue(counter)
-	if newValue > oldValue {
-		counter.Add(newValue - oldValue)
+	for labels, value := range mc.forwardedMetricsMap {
+		mc.updateCounterMetric(mc.forwardCountDesc, ch, value.count, labels.direction)
+		mc.updateCounterMetric(mc.forwardByteDesc, ch, value.bytes, labels.direction)
+	}
+
+	for labels, value := range mc.droppedMetricsMap {
+		mc.updateCounterMetric(mc.droppedCountDesc, ch, value.count, labels.reason, labels.direction)
+		mc.updateCounterMetric(mc.droppedByteDesc, ch, value.bytes, labels.reason, labels.direction)
 	}
 }
 
-// updatePrometheusMetrics checks the metricsmap key value pair
-// and determines which prometheus metrics along with respective labels
-// need to be updated.
-func updatePrometheusMetrics(key *Key, values *Values) {
-	updateMetric(func() (prometheus.Counter, error) {
-		if key.IsDrop() {
-			return metrics.DropCount.GetMetricWithLabelValues(key.DropForwardReason(), key.Direction())
-		}
-		return metrics.ForwardCount.GetMetricWithLabelValues(key.Direction())
-	}, float64(values.Count()))
+// Helper function that returns metricValues if key is present in map.
+// Otherwise, metricValues for key will be created and returned.
+func lookupOrCreateMetricsMapValues[K comparable](m map[K]metricValues, key K) metricValues {
+	if entry, ok := m[key]; ok {
+		return entry
+	} else {
```

Comment:
nit: superfluous else

Reply:
@derailed Thanks for review. I'll make changes after we resolve the issue with CI.